flink-formats/flink-csv/src/main/java/org/apache/flink/formats/csv/CsvToRowDataConverters.java

```diff
@@ -221,12 +221,22 @@ private int convertToDate(JsonNode jsonNode) {
 		return (int) Date.valueOf(jsonNode.asText()).toLocalDate().toEpochDay();
 	}
 
-	private int convertToTime(JsonNode jsonNode) {
+	private CsvToRowDataConverter convertToTime(TimeType timeType) {
+		final int precision = timeType.getPrecision();
 		// csv currently is using Time.valueOf() to parse time string
-		LocalTime localTime = Time.valueOf(jsonNode.asText()).toLocalTime();
 		// TODO: FLINK-17525 support millisecond and nanosecond
 		// get number of milliseconds of the day
-		return localTime.toSecondOfDay() * 1000;
+		return jsonNode -> {
+			LocalTime localTime = LocalTime.parse(jsonNode.asText());
+			if (precision == 3) {
+				return localTime.toNanoOfDay() / 1000_000L;
+			} else if (precision == 0) {
+				return localTime.toSecondOfDay() * 1000L;
+			} else {
+				throw new IllegalArgumentException("Csv does not support TIME type " +
+					"with precision: " + precision + ", it only supports precision 0 or 3.");
+			}
+		};
```

Comment:
```java
		final int precision = timeType.getPrecision();
		if (precision > 3) {
			throw new IllegalArgumentException("CSV format does not support TIME type " +
				"with precision: " + precision + ", it only supports precision 0 ~ 3.");
		}
		// get number of milliseconds of the day
		return jsonNode -> {
			LocalTime localTime = LocalTime.parse(jsonNode.asText());
			if (precision == 0) {
				return localTime.toSecondOfDay() * 1000L;
			} else {
				return (int) (localTime.toNanoOfDay() / 1000_000L);
			}
		};
```
Throws exception in compile phase instead during runtime. 
Besides, we can also support precision 0~3.

Reply:
I will optimize.